resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -640,34 +626,265 @@ protected boolean doIsResolved(InetSocketAddress address) {
         // InetSocketAddress.getHostString() is available since Java 7.
+        final String hostname;
         if (PlatformDependent.javaVersion() < 7) {
-            return addr.getHostName();
+            hostname = addr.getHostName();
         } else {
-            return addr.getHostString();
+            hostname = addr.getHostString();
+        }
+
+        return IDN.toASCII(hostname);
+    }
+
+    void cache(String hostname, InetAddress address, long originalTtl) {
+        final int maxTtl = maxTtl();
+        if (maxTtl == 0) {
+            return;
         }
+
+        final int ttl = Math.max(minTtl(), (int) Math.min(maxTtl, originalTtl));
+        final List<DnsCacheEntry> entries = cachedEntries(hostname);
+        final DnsCacheEntry e = new DnsCacheEntry(hostname, address);
+
+        synchronized (entries) {
+            if (!entries.isEmpty()) {
+                final DnsCacheEntry firstEntry = entries.get(0);
+                if (firstEntry.cause() != null) {
+                    assert entries.size() == 1;
+                    firstEntry.cancelExpiration();
+                    entries.clear();
+                }
+            }
+            entries.add(e);
+        }
+
+        scheduleCacheExpiration(entries, e, ttl);
+    }
+
+    void cache(String hostname, Throwable cause) {
+        final int negativeTtl = negativeTtl();
+        if (negativeTtl == 0) {
+            return;
+        }
+
+        final List<DnsCacheEntry> entries = cachedEntries(hostname);
+        final DnsCacheEntry e = new DnsCacheEntry(hostname, cause);
+
+        synchronized (entries) {
+            final int numEntries = entries.size();
+            for (int i = 0; i < numEntries; i ++) {
+                entries.get(i).cancelExpiration();
+            }
+            entries.clear();
+            entries.add(e);
+        }
+
+        scheduleCacheExpiration(entries, e, negativeTtl);
+    }
+
+    private List<DnsCacheEntry> cachedEntries(String hostname) {
+        List<DnsCacheEntry> oldEntries = resolveCache.get(hostname);
+        final List<DnsCacheEntry> entries;
+        if (oldEntries == null) {
+            List<DnsCacheEntry> newEntries = new ArrayList<DnsCacheEntry>();
+            oldEntries = resolveCache.putIfAbsent(hostname, newEntries);
+            entries = oldEntries != null? oldEntries : newEntries;
+        } else {
+            entries = oldEntries;
+        }
+        return entries;
+    }
+
+    private void scheduleCacheExpiration(final List<DnsCacheEntry> entries, final DnsCacheEntry e, int ttl) {
+        e.scheduleExpiration(
+                ch.eventLoop(),
+                new OneTimeTask() {
+                    @Override
+                    public void run() {
+                        synchronized (entries) {
+                            entries.remove(e);
+                            if (entries.isEmpty()) {
+                                resolveCache.remove(e.hostname());
```

Comment:
Let me just leave it as it is and fix it in a new PR that will include the cache rework because the race condition isn't critical (it just sometimes loses the cached entries.) and this PR was open for too long.